### PR TITLE
Add Edit Transaction Functionality

### PR DESCRIPTION
**Pull request by PetrichorPrecipice**
Add Edit Transaction Functionality 

Add classes EditTransactionCommandand EditTransactionCommandParser.

Add methods to parse type, description, amount, date and Person to
parseUtil.

Modify Transact class to take a Date parameter.

Add isValid methods to each parameter class (e.g. description, amount)
to help with parsing and their respective regexes.

Add Type enum.

Rename EditCommand to EditStaffCommand and EditCommandParser to
EditStaffCommandParser.

### CODE REVIEW

**Comment by PetrichorPrecipice** [comment]
Also updated transaction constructor so its method signature is now Transaction(transaction_id, type, description, amount, date, person)

File: src/main/java/transact/model/transaction/Expense.java
Comment (cxo05) [inline src/main/java/transact/model/transaction/Expense.java:1]: Delete this file

File: src/main/java/transact/model/transaction/Revenue.java
Comment (cxo05) [inline src/main/java/transact/model/transaction/Revenue.java:1]: Delete this file

**Review by cxo05 — CHANGES_REQUESTED** [review]

File: src/main/java/transact/model/transaction/Revenue.java
Comment (cxo05) [inline src/main/java/transact/model/transaction/Revenue.java:1]: Delete this file

File: src/main/java/transact/model/transaction/Expense.java
Comment (cxo05) [inline src/main/java/transact/model/transaction/Expense.java:1]: Delete this file

File: src/main/java/transact/model/transaction/info/Type.java
Comment (cxo05) [inline src/main/java/transact/model/transaction/info/Type.java:1]: Can rename to TransactionType?

Only got Revenue and Expense. Income/Profit is Revenue - Expense

Can use full word for the enums? (e.g. R => Revenue) but still allow shortform in isValidType() and getType()

File: src/main/java/transact/logic/commands/EditTransactionCommand.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+
+
+
+/**
+ * Edits the details of an existing person in the address book.
+ */
+public class EditTransactionCommand extends Command {
+
+    public static final String COMMAND_WORD = "edittransaction";
```
Comment (Bombbird2001) [inline src/main/java/transact/logic/commands/EditTransactionCommand.java:38]:
```suggestion
    public static final String COMMAND_WORD = "edit";
```
I think we can shorten this so the user doesn't have to type so much, same for addtransaction

File: src/main/java/transact/logic/commands/EditTransactionCommand.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            + "[" + PREFIX_DESCRIPTION + "DESCRIPTION] "
+            + "[" + PREFIX_AMOUNT + "AMOUNT] "
+            + "[" + PREFIX_DATE + "DATE] "
+            + "[" + PREFIX_STAFF + "STAFF]...\n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_TYPE + "E "
+            + PREFIX_AMOUNT + "10000";
+    public static final String MESSAGE_EDIT_TRANSACTION_SUCCESS = "Edited Transaction: %1$s";
+    public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
+    public static final String MESSAGE_DUPLICATE_PERSON = "This transaction already exists in the Transaction book.";
```
Comment (Bombbird2001) [inline src/main/java/transact/logic/commands/EditTransactionCommand.java:54]:
```suggestion
    public static final String MESSAGE_DUPLICATE_TRANSACTION = "This transaction already exists in the Transaction book.";
```

File: src/main/java/transact/logic/commands/EditTransactionCommand.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_TRANSACTION_DISPLAYED_INDEX);
+        }
+
+        Transaction transactionToEdit = lastShownList.get(index.getZeroBased());
+        Transaction editedTransaction = createEditedTransaction(transactionToEdit, editTransactionDescriptor);
+
+        if (!transactionToEdit.isSameEntry(editedTransaction) && model.hasTransaction(editedTransaction)) {
+            throw new CommandException(MESSAGE_DUPLICATE_PERSON);
```
Comment (Bombbird2001) [inline src/main/java/transact/logic/commands/EditTransactionCommand.java:86]:
```suggestion
            throw new CommandException(MESSAGE_DUPLICATE_TRANSACTION);
```

File: src/main/java/transact/logic/commands/EditTransactionCommand.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
+package transact.logic.commands;
+
+import static java.util.Objects.requireNonNull;
+import static transact.logic.parser.CliSyntax.PREFIX_AMOUNT;
+import static transact.logic.parser.CliSyntax.PREFIX_DATE;
+import static transact.logic.parser.CliSyntax.PREFIX_DESCRIPTION;
+import static transact.logic.parser.CliSyntax.PREFIX_STAFF;
+import static transact.logic.parser.CliSyntax.PREFIX_TYPE;
+import static transact.model.Model.PREDICATE_SHOW_ALL_PERSONS;
```
Comment (Bombbird2001) [inline src/main/java/transact/logic/commands/EditTransactionCommand.java:9]:
```suggestion
import static transact.model.Model.PREDICATE_SHOW_ALL_TRANSACTIONS;
```

File: src/main/java/transact/logic/commands/EditTransactionCommand.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+        Transaction transactionToEdit = lastShownList.get(index.getZeroBased());
+        Transaction editedTransaction = createEditedTransaction(transactionToEdit, editTransactionDescriptor);
+
+        if (!transactionToEdit.isSameEntry(editedTransaction) && model.hasTransaction(editedTransaction)) {
+            throw new CommandException(MESSAGE_DUPLICATE_PERSON);
+        }
+
+        model.setTransaction(transactionToEdit, editedTransaction);
+        model.updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
```
Comment (Bombbird2001) [inline src/main/java/transact/logic/commands/EditTransactionCommand.java:90]:
```suggestion
        model.updateFilteredTransactionList(PREDICATE_SHOW_ALL_TRANSACTIONS);
```

File: src/main/java/transact/logic/commands/EditTransactionCommand.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .add("index", index)
+                .add("editTransactionDescriptor", editTransactionDescriptor)
+                .toString();
+    }
+
+    /**
+     * Stores the details to edit the person with. Each non-empty field value will
```
Comment (Bombbird2001) [inline src/main/java/transact/logic/commands/EditTransactionCommand.java:136]:
```suggestion
     * Stores the details to edit the transaction with. Each non-empty field value will
```

File: src/main/java/transact/logic/commands/EditTransactionCommand.java (hunk 1 above)
Comment (cxo05) [reply src/main/java/transact/logic/commands/EditTransactionCommand.java:38]: Mmm maybe can `edit t` or `edit s` like `view`

**Comment by cxo05** [comment]
You changed all instances of "type" in the project to TransactionType lul

File: src/main/resources/view/MainWindow.fxml
Comment (cxo05) [inline src/main/resources/view/MainWindow.fxml:1]: Revert

File: src/main/java/transact/ui/TransactionTablePanel.java
Comment (cxo05) [inline src/main/java/transact/ui/TransactionTablePanel.java:1]: Revert

File: docs/_sass/minima/_layout.scss
Comment (cxo05) [inline docs/_sass/minima/_layout.scss:1]: Revert

File: docs/_includes/header.html
Comment (cxo05) [inline docs/_includes/header.html:1]: Revert

File: config/checkstyle/checkstyle.xml
Comment (cxo05) [inline config/checkstyle/checkstyle.xml:1]: Revert

File: docs/DeveloperGuide.md
Comment (cxo05) [inline docs/DeveloperGuide.md:1]: Revert

**Pull request merged**